lib/auth/native/native.go

```diff
@@ -133,35 +134,56 @@ func (n *nauth) GenerateKeyPair(passphrase string) ([]byte, []byte, error) {
 	return privPem, pubBytes, nil
 }
 
-func (n *nauth) GenerateHostCert(privateSigningKey, publicKey []byte, hostname, authDomain string, roles teleport.Roles, ttl time.Duration) ([]byte, error) {
-	if err := roles.Check(); err != nil {
+func (n *nauth) GenerateHostCert(c services.CertParams) ([]byte, error) {
+	if err := c.Check(); err != nil {
 		return nil, trace.Wrap(err)
 	}
-	pubKey, _, _, _, err := ssh.ParseAuthorizedKey(publicKey)
+
+	pubKey, _, _, _, err := ssh.ParseAuthorizedKey(c.PublicHostKey)
 	if err != nil {
-		return nil, err
+		return nil, trace.Wrap(err)
+	}
+
+	signer, err := ssh.ParsePrivateKey(c.PrivateCASigningKey)
+	if err != nil {
+		return nil, trace.Wrap(err)
 	}
+
+	// create a slice of principals to add to the certificate. the values
+	// for the principals need to be the fully qualified name that will be
+	// used when connecting to them which is why they also contain a domain
+	// name (cluster name).
+	//
+	// two principals are always added, the first is uuid.clusterName and
+	// the second is nodeName.clusterName. The first is for tsh use internally,
+	// while the second is for openssh interoperability where people will be
+	// connecting by the dns name.
+	principals := []string{
+		fmt.Sprintf("%s.%s", c.HostID, c.ClusterName),
```

Comment:
what if `HostID` or `NodeName` are passed empty? you have to check parameters for sanity I think